backend/llm/openai.py

```diff
@@ -148,46 +131,63 @@ def _create_llm(self, model_name, streaming=False, callbacks=None) -> LLM:
             callbacks=callbacks,
         )
 
-    def _get_qa(
-        self,
-    ) -> ConversationalRetrievalChain:
+    def _call_chain(self, chain, question, history):
         """
-        Retrieves a QA chain for the given chat message and API key.
-        :param chat_message: The chat message containing history.
-        :param user_openai_api_key: The OpenAI API key to be used.
-        :return: ConversationalRetrievalChain instance
+        Call a chain with a given question and history.
+        :param chain: The chain eg QA (ConversationalRetrievalChain)
+        :param question: The user prompt
+        :param history: The chat history from DB
+        :return: The answer.
         """
-
-        # Initialize and return a ConversationalRetrievalChain
-        qa = ConversationalRetrievalChain(
-            retriever=self.vector_store.as_retriever(),
-            question_generator=self.question_generator,
-            combine_docs_chain=self.doc_chain,
-            verbose=True,
+        return chain(
+            {
+                "question": question,
+                "chat_history": history,
+            }
         )
 
-        return qa
-
-    def generate_answer(self, question: str) -> str:
+    def generate_answer(self, question: str) -> ChatHistory:
         """
         Generate an answer to a given question by interacting with the language model.
         :param question: The question
         :return: The generated answer.
         """
         transformed_history = []
 
-        # Get the QA chain
-        qa = self._get_qa()
+        # Get the history from the database
         history = get_chat_history(self.chat_id)
 
         # Format the chat history into a list of tuples (human, ai)
-        transformed_history = [(chat.user_message, chat.assistant) for chat in history]
+        transformed_history = self.format_chat_history(history)
 
         # Generate the model response using the QA chain
-        model_response = qa({"question": question, "chat_history": transformed_history})
+        model_response = self._call_chain(self.qa, question, transformed_history)
```

Comment:
self.qa refers to `chain`. Pretty confused about the wording qa vs chain

Reply:
qa is the specific chain. _call_chain would work on any chain. Say we wanted to change to a summarize chain it would be a very simple change.